tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -79,77 +128,295 @@ private void OkButton_Click(object sender, RoutedEventArgs e)
+            var percentComplete = ShortcutProcessingEndIndex + ((j + 1) * ShortcutProcessingEndIndex / _itemCount);
+            progress.Report(percentComplete);
+        }
+
+        _allApps = _allApps.OrderBy(app => app.Name).ToList();
+
+        // We get the icon data on the UI thread, because BitmapImages must be created on the UI thread.
+        for (var k = 0; k < _allApps.Count; k++)
+        {
+            var app = _allApps[k];
+            if (app.IsMsix)
+            {
+                if (app.AppPackage is not null)
+                {
+                    try
+                    {
+                        // The package might be in a bad state, and accessing its
+                        // properties might throw an exception.
+                        app.Icon = new BitmapImage(app.AppPackage.Logo);
+                        app.IconFilePath = app.AppPackage.Logo.LocalPath;
+                    }
+                    catch
+                    {
+                        _log.Error("Error getting icon from package");
+                    }
+                }
+            }
+            else
+            {
+                if (!string.IsNullOrEmpty(app.TargetPath))
+                {
+                    app.Icon = WindowHelper.GetBitmapImageFromFile(app.TargetPath);
+                }
+            }
+
+            currentCount++;
+            var percentComplete = PackageProcessingEndIndex + ((k + 1) * ShortcutProcessingEndIndex / _itemCount);
+            progress.Report(percentComplete);
+
+            // Yield to make sure the UI thread can update the progress output.
+            await Task.Delay(1);
+        }
     }
 
-    private void UpdateSampleCommandline()
+    public void ProcessShortcut(string filePath)
     {
-        if (SampleCommandTextBox is null)
+        var appName = Path.GetFileNameWithoutExtension(filePath);
+
+        // Exclude Microsoft Virtual Desktop _shortcuts.
+        if (appName.Contains("Microsoft Virtual Desktop", StringComparison.OrdinalIgnoreCase))
         {
-            // The window is still initializing.
             return;
         }
 
-        var tool = GetCurrentToolDefinition();
-        if (tool is null)
+        var wshShell = new WshShell();
+        if (wshShell.CreateShortcut(filePath) is not IWshShortcut shortcut)
         {
+            _log.Error("Error getting shortcut");
             return;
         }
 
-        SampleCommandTextBox.Text = tool.CreateFullCommandLine(123, (Windows.Win32.Foundation.HWND)123);
+        // Proceed with using the shortcut object.
+        var targetPath = shortcut.TargetPath;
+
+        // Exclude _shortcuts that point to empty targets, filesystem folders, or WSA apps.
+        if (string.IsNullOrEmpty(targetPath)
+            || Directory.Exists(targetPath)
+            || targetPath.Contains("WsaClient.exe", StringComparison.OrdinalIgnoreCase))
+        {
+            return;
+        }
+
+        // Exclude *.chm, *.url, *.html, *.ico targets.
```

Comment:
>Exclude *.chm, *.url, *.html, *.ico targets. [](http://example.com/codeflow?start=11&length=44)

Presumably this list could be infinite. Do we only want to allow certain extensions instead (exe, bat, cmd, ps1, etc)?

Reply:
I filed a separate item for extension types - covers both the browse scenario and the shortcut scenario. Probably an inclusion rather than an exclusion list - but even with inclusion there's extra work to do (eg for ps1).